### PR TITLE
[DOC] How to test and write benchmarks

### PR DESCRIPTION
**Pull request by hmaarrfk**
## Description

@scikit-image/core please feel free to made modification by pushing to this repo. Hopefully this explains the 3 commands I use to write benchmarks well enough for it to get included in scikit-image.

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: CONTRIBUTING.txt [hunk 1]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+      def setup(self):
+          self.image = np.zeros((2000, 2000))
+          idx = np.arange(500, 1500)
+          self.image[idx[::-1], idx] = 255
+          self.image[idx, idx] = 255
+
+      def time_hough_line(self):
+          result1, result2, result3 = transform.hough_line(self.image)
+
+Here, the creation of the image is completed is in the ``setup`` method, and not
```
Comment (sciunto) [inline CONTRIBUTING.txt]: I do not understand this sentence, I guess there is a typo.

File: CONTRIBUTING.txt [hunk 2]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+Where the ``TransformSuite`` above will be run once in your current environment
+to test that everything is in order.
+
+Running your benchmark
+~~~~~~~~~~~~~~~~~~~~~~
+
+The command above is fast, but doesn't always test the performance of the code
+adequately. To do that you may want to run the benchmark in your current
+environment to see the performance of your change as you are developing new
+features. the command ``asv run -E existing`` will specify that you wish to run
```
Comment (sciunto) [inline CONTRIBUTING.txt]:
```suggestion
features. The command ``asv run -E existing`` will specify that you wish to run
```

File: README.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 - **Website (including documentation):** [https://scikit-image.org/](https://scikit-image.org)
 - **Mailing list:** [https://mail.python.org/mailman/listinfo/scikit-image](https://mail.python.org/mailman/listinfo/scikit-image)
 - **Source:** [https://github.com/scikit-image/scikit-image](https://github.com/scikit-image/scikit-image)
+- **Benchmarks:** [https://pandas.pydata.org/speed/scikit-image/](https://pandas.pydata.org/speed/scikit-image/)
```
Comment (sciunto) [inline README.md]: I would like to see this link in the CONTRIBUTING.txt file as well.

File: CONTRIBUTING.txt (hunk 1 above)
Comment (hmaarrfk) [reply CONTRIBUTING.txt]: I just want to bring to the attention of the contributor that things like creating our loading the dataset should not happen in the time function.

File: CONTRIBUTING.txt [hunk 4]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+To let airspeed velocity know more information about your machine.
+
+Writing a benchmark
+~~~~~~~~~~~~~~~~~~~
+To write  benchmark, add a file in the ``benchmarks`` directory which contains a
+a class with a ``setup`` method and a ``time_`` method.
+
+The setup method should contain all the code necessary to prepare your data
+before the function you wish to benchmark. The ``time_`` method
+should contain calls to the functions you wish to benchmark.
```
Comment (lagru) [inline CONTRIBUTING.txt]: I like your structure but I think this part could be more explicit. Maybe something like
```
The ``time_`` method should only contain code you wish to benchmark. 
Therefore it is useful to move everything that prepares the benchmark scenario
into the ``setup`` method. This function is called before calling a ``time_``
method and its execution time is not factored into the benchmarks.
```

File: CONTRIBUTING.txt [hunk 5]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+is more appropriate. Once installed, it is useful to run the command::
+
+  asv machine
+
+To let airspeed velocity know more information about your machine.
+
+Writing a benchmark
+~~~~~~~~~~~~~~~~~~~
+To write  benchmark, add a file in the ``benchmarks`` directory which contains a
+a class with a ``setup`` method and a ``time_`` method.
```
Comment (lagru) [inline CONTRIBUTING.txt]:
```suggestion
a class with one ``setup`` method and at least one method prefixed with ``time_``.
```

File: CONTRIBUTING.txt [hunk 6]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+          self.image[idx, idx] = 255
+
+      def time_hough_line(self):
+          result1, result2, result3 = transform.hough_line(self.image)
+
+Here, the creation of the image is completed is in the ``setup`` method, and not
+included in the reported time of the benchmark.
+
+It is also possible to benchmark features such as peak memory usage. To learn
+more about features of `asv`, we refer the read to the official
```
Comment (lagru) [inline CONTRIBUTING.txt]:
```suggestion
more about the features of `asv`, please refer to the official
```

File: CONTRIBUTING.txt [hunk 7]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+`airpseed velocity documentation <http://asv.readthedocs.io/en/latest/writing_benchmarks.html>`_.
+
+Testing the benchmarks locally
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Prior to running the true benchmark, it is often worthwhile to test that the
+code if free of typos. To do so, you may use the ``asv dev``::
```
Comment (lagru) [inline CONTRIBUTING.txt]:
```suggestion
code is free of typos. To do so, you may use the ``asv dev``::
```

File: CONTRIBUTING.txt [hunk 8]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+
+  asv dev -b TransformSuite
+
+Where the ``TransformSuite`` above will be run once in your current environment
+to test that everything is in order.
+
+Running your benchmark
+~~~~~~~~~~~~~~~~~~~~~~
+
+The command above is fast, but doesn't always test the performance of the code
```
Comment (lagru) [inline CONTRIBUTING.txt]:
```suggestion
The command above is fast, but doesn't test the performance of the code
```
I don't think the ``dev`` command is ever appropriate to benchmark.

File: CONTRIBUTING.txt [hunk 9]
```diff
@@ -511,3 +511,121 @@ Bugs
...
+of time since building scikit-image can be a time consuming task::
+
+  asv run -E existing -b TransformSuite
+
+Comparing results to master
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Often, the goal of a PR is to compare the results of the modifications in terms
+speed to a snapshot of the code that is in the master branch of the
+``scikit-image`` repository. The ``asv continuous`` is of help here::
```
Comment (lagru) [inline CONTRIBUTING.txt]:
```suggestion
``scikit-image`` repository. The command ``asv continuous`` is of help here::
```

File: CONTRIBUTING.txt (hunk 1 above)
Comment (lagru) [reply CONTRIBUTING.txt]: That's a good thought but may be redundant with my suggestion above. 
```suggestion
Here, the creation of the image is completed in the ``setup`` method, and not
```

**Comment by hmaarrfk** [comment]
I didn't even know about that section! I'll probably use that soon. Not too sure I want to explode this section with ASV specifics. Maybe cross linking to that section might help? I haven't seen many issues with inconsistencies with my laptop.

I mostly get tired of waiting, and so I start using my browser (which on linux doesn't have hardware decoding for video) so that messes with things alot.

File: CONTRIBUTING.txt (hunk 1 above)
Comment (sciunto) [reply CONTRIBUTING.txt]: I prefer @lagru suggestion then. It's easier to understand for me.

Comment (sciunto) [reply CONTRIBUTING.txt]: Then, it will be ready to be merged on my side.

File: CONTRIBUTING.txt (hunk 4 above)
Comment (hmaarrfk) [reply CONTRIBUTING.txt]: Thanks!

**Comment by jni** [comment]
This is wonderful! Thanks @hmaarrfk and thanks @lagru and @sciunto for the reviews!

**Pull request merged**